eras/shelley/impl/src/Cardano/Ledger/Shelley/TxWits.hs

```diff
@@ -119,27 +120,58 @@ deriving via
     (Era era, NoThunks (Script era)) =>
     (NoThunks (WitnessSetHKD Identity era))
 
-type ShelleyTxWits = WitnessSetHKD Identity
+newtype ShelleyTxWits era = ShelleyTxWitsConstr (WitnessSetHKD Identity era)
+
+deriving newtype instance EraScript era => Eq (ShelleyTxWits era)
+
+deriving newtype instance EraScript era => Show (ShelleyTxWits era)
+
+deriving newtype instance Era era => Generic (ShelleyTxWits era)
+
+instance
+  ( Era era,
+    NFData (Script era),
+    NFData (WitVKey 'Witness (EraCrypto era)),
+    NFData (BootstrapWitness (EraCrypto era))
+  ) =>
+  NFData (ShelleyTxWits era)
+
+instance (Era era, NoThunks (Script era)) => NoThunks (ShelleyTxWits era)
+
+-- =======================================================
+-- Accessors
+-- =======================================================
 
 -- | Addresses witness setter and getter for `ShelleyTxWits`. The
 -- setter does update memoized binary representation.
 addrShelleyWitsL ::
-  EraTxWits era => Lens' (ShelleyTxWits era) (Set (WitVKey 'Witness (EraCrypto era)))
-addrShelleyWitsL = lens addrWits' (\w s -> w {addrWits = s})
+  EraScript era => Lens' (ShelleyTxWits era) (Set (WitVKey 'Witness (EraCrypto era)))
+addrShelleyWitsL =
+  lens
+    (\(ShelleyTxWitsConstr w) -> addrWits' w)
+    (\(ShelleyTxWitsConstr w) s -> ShelleyTxWits {addrWits = s, scriptWits = scriptWits' w, bootWits = bootWits' w})
 {-# INLINEABLE addrShelleyWitsL #-}
 
 -- | Bootstrap Addresses witness setter and getter for `ShelleyTxWits`. The
 -- setter does update memoized binary representation.
 bootAddrShelleyWitsL ::
-  EraTxWits era => Lens' (ShelleyTxWits era) (Set (BootstrapWitness (EraCrypto era)))
-bootAddrShelleyWitsL = lens bootWits' (\w s -> w {bootWits = s})
+  EraScript era =>
+  Lens' (ShelleyTxWits era) (Set (BootstrapWitness (EraCrypto era)))
+bootAddrShelleyWitsL =
+  lens
+    (\(ShelleyTxWitsConstr w) -> bootWits' w)
+    (\(ShelleyTxWitsConstr w) s -> ShelleyTxWits {addrWits = addrWits' w, scriptWits = scriptWits' w, bootWits = s})
```

Comment:
@lehins In the end, the line with the setter that you showed me  (` ShelleyTxWits (w {addrWits = s}))` ) didn't work for me exactly like that, I think because the definition of the pattern was different.  Or maybe I messed something else up! 

If you know of a better way let me know please

Reply:
This should work:
```haskell
bootAddrShelleyWitsL =
  lens
    (\(ShelleyTxWitsConstr w) -> bootWits' w)
    (\(ShelleyTxWitsConstr w) s -> ShelleyTxWitsConstr (w {bootWits = s}))
```

Reply:
That's what I tried, but with the pattern definition like this: 

```pattern ShelleyTxWits ::
  EraScript era =>
  Set (WitVKey 'Witness (EraCrypto era)) ->
  Map (ScriptHash (EraCrypto era)) (Script era) ->
  Set (BootstrapWitness (EraCrypto era)) ->
  ShelleyTxWits era
pattern ShelleyTxWits {addrWits, scriptWits, bootWits} <- 
...
```

I'm getting 

```
    • Couldn't match expected type ‘WitnessSetHKD Identity era’
                  with actual type ‘ShelleyTxWits era0’
    • In the first argument of ‘ShelleyTxWitsConstr’, namely
        ‘(w {bootWits = s})’
```

Maybe the pattern should be defined in another way? :thinking:

Reply:
Oh yeah, you've changed already the type of `ShelleTxWits` pattern synonym, so it should just be:
```haskell
bootAddrShelleyWitsL =
  lens
    (\(ShelleyTxWitsConstr w) -> bootWits' w)
    (\w s -> w {bootWits = s})
```

Reply:
Oh! :facepalm: